pkg/api/handlers/compat/images_build.go

```diff
@@ -169,14 +170,60 @@ func BuildImage(w http.ResponseWriter, r *http.Request) {
 
 	// convert addcaps formats
 	containerFiles := []string{}
-	if _, found := r.URL.Query()["dockerfile"]; found {
-		var m = []string{}
-		if err := json.Unmarshal([]byte(query.Dockerfile), &m); err != nil {
-			// it's not json, assume just a string
-			m = []string{filepath.Join(contextDirectory, query.Dockerfile)}
+	// If query paramemter `dockerfile` is set and is
+	// on priority never look for `Dockerfile` or
+	// `Containerfile` in the context directory.
+	dockerFileSet := false
+	// Query paramemter `remote` always takes priority
+	// over parameter `dockerfile`.
+	remoteSet := false
+	if utils.IsLibpodRequest(r) && query.Remote != "" {
+		// The context directory could be a URL.  Try to handle that.
+		anchorDir, err := ioutil.TempDir("", "libpod_builder")
+		if err != nil {
+			utils.BadRequest(w, "remote", query.Remote, err)
 		}
-		containerFiles = m
-	} else {
+		tempDir, subDir, err := buildahDefine.TempDirForURL(anchorDir, "buildah", query.Remote)
+		if err != nil {
+			utils.BadRequest(w, "remote", query.Remote, err)
+		}
+		if tempDir != "" {
+			// We had to download it to a temporary directory.
+			// Delete it later.
+			defer func() {
+				if err = os.RemoveAll(tempDir); err != nil {
+					// We are deleting this on server so log on server end
+					// client does not have to worry about server cleanup.
+					logrus.Errorf("Cannot delete downloaded temp file %q: %s", tempDir, err)
+				}
+			}()
+			contextDirectory = filepath.Join(tempDir, subDir)
+		} else {
+			// Nope, it was local.  Use it as is.
+			absDir, err := filepath.Abs(query.Remote)
+			if err != nil {
+				utils.BadRequest(w, "remote", query.Remote, err)
+			}
+			contextDirectory = absDir
+		}
+		remoteSet = true
+	}
+
+	// Allow `dockerfile` only if `remote` was not configured
+	// since `remote` overrides `dockerfile` as per API docs.
+	if !remoteSet {
```

Comment:
Can't this just be an else for previous block and not need for remoteSet variable?

Reply:
Yes can be merged to else block.